### PR TITLE
JCSV Module Implemented

### PR DESCRIPTION
**Pull request by suleman01851**
Assalam o Alaikum, Faizan Bhai!
With my limited knowledge of GOlang, I tried my best to manage this module, but could not complete it. I think I should opt for [ ]byte and a boolean as data members, to determine whether it is csv or json, and convert it later on. Kindly review this, and enlighten me how it should be programmed.
@IamFaizanKhalid 
@fahadsiddiqui

### CODE REVIEW

File: file.go [hunk 1]
```diff
@@ -1,37 +1,115 @@
-package jcsv
+package main
```
Comment (IamFaizanKhalid) [inline file.go]: Why have you changed the package `jcsv` to `main`?

File: file.go [hunk 2]
```diff
@@ -1,37 +1,115 @@
...
+	"encoding/csv"
+	"fmt"
+	"io/ioutil"
+	"os"
+)
 
 type file struct {
-	// TODO: define variables if needed
+	data []uint8
+	csvData [][]string
```
Comment (IamFaizanKhalid) [inline file.go]: Reading this code, I can't understand what is the difference between these two. There are no comments either.

File: file.go [hunk 3]
```diff
@@ -1,37 +1,115 @@
...
+	csvHeader []string
 }
 
 func ParseJsonFile(path string) (file, error) {
-	// TODO: open and read the given file into your `file` object
-	return file{}, nil
+	var f file
+	f.data=nil
+	fileStream,fileOpenError := os.Open(path)
```
Comment (IamFaizanKhalid) [inline file.go]: Please read this for naming conventions.
Note: These are just conventions, not hard rules.

File: file.go [hunk 4]
```diff
@@ -1,37 +1,115 @@
...
+	data []uint8
+	csvData [][]string
+	csvHeader []string
 }
 
 func ParseJsonFile(path string) (file, error) {
-	// TODO: open and read the given file into your `file` object
-	return file{}, nil
+	var f file
+	f.data=nil
```
Comment (IamFaizanKhalid) [inline file.go]: Why do we need to do this?

File: file.go [hunk 5]
```diff
@@ -1,37 +1,115 @@
...
+	fileStream,fileOpenError := os.Open(path)
+	if fileOpenError!=nil{
+		return file{},fileOpenError
+	}
+	var fileReadError error
```
Comment (IamFaizanKhalid) [inline file.go]: Add an empty line above this. Just to give your code a clean look.

Comment (IamFaizanKhalid) [reply file.go]: Is it necessary to declare it here separately?

File: file.go [hunk 6]
```diff
@@ -1,37 +1,115 @@
...
+	f.data,fileReadError=ioutil.ReadAll(fileStream)
+	if fileReadError!=nil {
+		return file{},fileReadError
+	}
+	return f,nil
```
Comment (IamFaizanKhalid) [inline file.go]: The code is not formatted properly.
Use the command `go fmt ./...` in your project root directory to format your whole project.

File: file.go [hunk 7]
```diff
@@ -1,37 +1,115 @@
...
 }
 
 func ParseOpenedJsonFile(f *os.File) (file, error) {
-	// TODO: read the given file into your `file` object
-	return file{}, nil
+	var fileRead file
+	var fileReadError error
+	fileRead.data,fileReadError=ioutil.ReadAll(f)
```
Comment (IamFaizanKhalid) [inline file.go]:
```suggestion
	fileRead.data, fileReadError := ioutil.ReadAll(f)
```
Can also be done like this.

Comment (IamFaizanKhalid) [reply file.go]: `fileRead.data` is a `[]uint8`. Are you sure this code works?

File: file.go [hunk 8]
```diff
@@ -1,37 +1,115 @@
...
+	return fileRead,nil
 }
 
 func ParseCsvFile(path string, hasHeaders bool) (file, error) {
-	// TODO: open and read the given file into your `file` object
-	return file{}, nil
+	var f file
+	f.data=nil
+	f.csvData=nil
+	f.csvHeader=nil
```
Comment (IamFaizanKhalid) [inline file.go]: Why do we have to do this?

File: file.go [hunk 9]
```diff
@@ -1,37 +1,115 @@
...
+	if err != nil {
+		fmt.Println(err)
+		return file{},err
+	}
+	readBuffer,readError := csv.NewReader(csvFile).ReadAll()
+	if readError!=nil{
+		return file{},readError
+	}
+	if hasHeaders{
+		for j:=0;j<len(readBuffer[0]);j++{
```
Comment (IamFaizanKhalid) [inline file.go]: What will happen if the csv file is empty and there are no rows. Will `readBuffer[0]` work?

File: file.go [hunk 10]
```diff
@@ -1,37 +1,115 @@
...
+			f.csvHeader=append(f.csvHeader,readBuffer[0][j])
+		}
+	}
+	var i int
+	if hasHeaders{
+		i=1
+	}else{
+		i=0
+	}
```
Comment (IamFaizanKhalid) [inline file.go]: Can't we do this where you checked `if hasHeaders` before.

File: file.go [hunk 11]
```diff
@@ -1,37 +1,115 @@
...
+	for ;i<len(readBuffer);i++{
+		f.csvData = append(f.csvData, readBuffer[i])
+	}
```
Comment (IamFaizanKhalid) [inline file.go]: `readBuffer` is `[][]string`, and `f.csvData` is also `[][]string`.
Can't we do something like `f.csvData=readBuffer`?

File: file.go [hunk 12]
```diff
@@ -1,37 +1,115 @@
...
 func (f file) Csv(addHeaders bool) []byte {
-	// TODO: return the file data in CSV format
-	return nil
+	var data []byte
+	for i:=0;i< len(f.csvData);i++{
+		for j:=0;j<len(f.csvData[i]);j++{
+			for k:=0;k<len(f.csvData[i][j]);k++{
+				data = append(data , f.csvData[i][j][k])
+			}
+			data = append(data, ',')
```
Comment (IamFaizanKhalid) [inline file.go]: Can't we use `strings.Join()` instead of this?

File: file.go [hunk 13]
```diff
@@ -1,37 +1,115 @@
...
 }
 
 func (f file) Csv(addHeaders bool) []byte {
-	// TODO: return the file data in CSV format
-	return nil
+	var data []byte
+	for i:=0;i< len(f.csvData);i++{
+		for j:=0;j<len(f.csvData[i]);j++{
+			for k:=0;k<len(f.csvData[i][j]);k++{
+				data = append(data , f.csvData[i][j][k])
```
Comment (IamFaizanKhalid) [inline file.go]: Isn't it better to use loop with `range` instead of doing `f.csvData[i][j][k]`?

Comment (IamFaizanKhalid) [reply file.go]: `f.csvData` is of type `[][]string`.
`f.csvData[i]`  is of type `[]string`.
`f.csvData[i][j]`  is of type `string`.
Why don't you append a whole string instead of appending character by character using `f.csvData[i][j][k]`?

File: file.go [hunk 14]
```diff
@@ -1,37 +1,115 @@
...
+		data=append(data,'\n')
+	}
+	return data
+
 }
 
 func (f file) Json() []byte {
-	// TODO: return the file data in JSON format
-	return nil
+	return f.data
```
Comment (IamFaizanKhalid) [inline file.go]: When did you convert `csv` file into JSON format?
If I do `jcsv.ParseCsvFile("input.csv").Json()`, will it give me the proper json data?

**Review by IamFaizanKhalid — CHANGES_REQUESTED** [review]
Not completing the review, because you haven't tested it properly yourself. Please ask for a review when you have tested it and it is working properly.

File: file.go (hunk 3 above)
Comment (IamFaizanKhalid) [reply file.go]: You are opening the file but not closing it.

File: file.go (hunk 1 above)
Comment (suleman01851) [reply file.go]: GoLand editor was prompting an error for not finding `main` package.

> Why have you changed the package `jcsv` to `main`?

File: file.go (hunk 4 above)
Comment (suleman01851) [reply file.go]: > Why do we need to do this?

Yes you are right, GO automatically assigns them nil.

File: file.go (hunk 5 above)
Comment (suleman01851) [reply file.go]: > Is it necessary to declare it here separately?

I had to, because I had already declared `file` object, and on the same statement, I couldn't initialize a variable and define a new variable at the same time

File: file.go (hunk 11 above)
Comment (suleman01851) [reply file.go]: > `readBuffer` is `[][]string`, and `f.csvData` is also `[][]string`. Can't we do something like `f.csvData=readBuffer`?

Right now, I am not sure, assignment operator does shallow copy or deep, to be on the safe side, I tried doing this through loop.

File: file.go (hunk 13 above)
Comment (suleman01851) [reply file.go]: > `f.csvData` is of type `[][]string`. `f.csvData[i]` is of type `[]string`. `f.csvData[i][j]` is of type `string`. Why don't you append a whole string instead of appending character by character using `f.csvData[i][j][k]`?

I am not sure why I used third loop, I will rectify it.

File: file.go (hunk 1 above)
Comment (IamFaizanKhalid) [reply file.go]: That is because you have a `main()` function to execute. `main()` function should be in the `main` package.
Don't add `main()` in the package `jscv`.

File: file.go (hunk 11 above)
Comment (IamFaizanKhalid) [reply file.go]: For array, it copies its values.
For slices, it copies its references.

https://stackoverflow.com/questions/25083235/go-assign-multiple-return-value-function-to-new-and-old-variable

Comment (IamFaizanKhalid) [reply file.go]: There is also a built-in function in golang for the `copy()`. You can also check how it works. 
https://pkg.go.dev/builtin#copy

File: jcsv.go [hunk 15]
```diff
@@ -9,3 +14,62 @@ func CsvToJson(c []byte, hasHeaders bool) ([]byte, error) {
 	// TODO: convert CSV data in `c` into JSON format and return
 	return nil, nil
 }
+
+func main() {
```
Comment (IamFaizanKhalid) [inline jcsv.go]: There shouldn't be a `main()` function in the package `jcsv`. You can add this for testing, but don't commit it. For testing you can write proper `_test.go` files.

Comment (suleman01851) [reply jcsv.go]: Sure, I will follow this paradigm from now on.

File: file.go (hunk 13 above)
Comment (suleman01851) [reply file.go]: > `f.csvData` is of type `[][]string`. `f.csvData[i]` is of type `[]string`. `f.csvData[i][j]` is of type `string`. Why don't you append a whole string instead of appending character by character using `f.csvData[i][j][k]`?

I tried doing it this way, it prompts error for appending a string to [] byte.

File: file.go [hunk 16]
```diff
@@ -1,37 +1,127 @@
...
 func ParseJsonFile(path string) (file, error) {
-	// TODO: open and read the given file into your `file` object
-	return file{}, nil
+	var fileObj file
+	fileStream,fileOpenError := os.Open(path)
+	if fileOpenError!=nil{
+		return file{},fileOpenError
+	}
+	var fileReadError error
+	fileObj.jsonData,fileReadError=ioutil.ReadAll(fileStream)
```
Comment (IamFaizanKhalid) [inline file.go]: Please keep an empty line between each code block to make your code look cleaner.

File: file.go [hunk 17]
```diff
@@ -1,37 +1,127 @@
...
+	fileStream.Close()
```
Comment (IamFaizanKhalid) [inline file.go]: Please read about `defer`.

File: file.go (hunk 16 above)
Comment (IamFaizanKhalid) [inline file.go]: Does assignment to `fileObj.jsonData` works?
Have you tested it for whether it causes a null pointer dereference error or not?

File: file.go [hunk 18]
```diff
@@ -1,37 +1,127 @@
...
+	"encoding/csv"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+)
 
 type file struct {
-	// TODO: define variables if needed
+	jsonData []uint8
```
Comment (IamFaizanKhalid) [inline file.go]: What is the reason for using `[]uint8`?

File: file.go [hunk 19]
```diff
@@ -1,37 +1,127 @@
...
+		return file{},err
+	}
+	readBuffer,readError := csv.NewReader(csvFile).ReadAll()
+	if readError!=nil{
+		return file{},readError
+	}
+	var i int
+	if hasHeaders {
+		for j := 0; j < len(readBuffer[0]); j++ {
+			returnFile.csvHeader = append(returnFile.csvHeader, readBuffer[0][j])
```
Comment (IamFaizanKhalid) [inline file.go]: `returnFile.csvHeader` is an `[]string`.
`readBuffer[0]` too is an `[]string`.
Please check if we can do the following or not:
```go
    returnFile.csvHeader = readBuffer[0]
```

File: file.go [hunk 20]
```diff
@@ -1,37 +1,127 @@
...
+		}
+		i=1
+	}else {
```
Comment (IamFaizanKhalid) [inline file.go]: Code is not formatted.

File: file.go [hunk 21]
```diff
@@ -1,37 +1,127 @@
...
+	}
+	return  fileObj,nil
 }
 
 func (f file) Csv(addHeaders bool) []byte {
-	// TODO: return the file data in CSV format
-	return nil
+	var dataToBeReturned []byte
+	if addHeaders == true{
+		for j:=0;j<len(f.csvHeader);j++{
```
Comment (IamFaizanKhalid) [inline file.go]: If the input file was a JSON file, this will fail.

**Review by IamFaizanKhalid — CHANGES_REQUESTED** [review]
Please test your code with the following before requesting a review:

```go
    jcsv.ParseCsvFile("input.csv").Json()
```
```go
    jcsv.ParseJsonFile("input.json").Csv()
```

File: jcsv.go [hunk 22]
```diff
@@ -1,11 +1,107 @@
...
+	var result string
+	i++
+	count := 0
+	for ; i < len(c); i++ {
+		for ; c[i] != '\n'; i++ {
+			if c[i] == ',' {
+				if count == 0 {
+					result += "{\n"
+				}
+				result += string("\t\"") + attributes[count] + string("\":\"") + strings.TrimLeft(string(c[begin:i]), "\n") + string("\",\n")
```
Comment (IamFaizanKhalid) [inline jcsv.go]: You are converting a string to a string.

File: jcsv.go [hunk 23]
```diff
@@ -1,11 +1,107 @@
...
 
 func CsvToJson(c []byte, hasHeaders bool) ([]byte, error) {
-	// TODO: convert CSV data in `c` into JSON format and return
-	return nil, nil
+	if c == nil {
+		return nil, nil
+	}
+	var attributes []string
+	begin := 0
+	keysCount := 1
```
Comment (IamFaizanKhalid) [inline jcsv.go]: There are no comments to understand what these variables are for. Add some comments, please.

**Review by IamFaizanKhalid — CHANGES_REQUESTED** [review]

File: file.go (hunk 12 above)
Comment (suleman01851) [reply file.go]: Here I am converting Csv data, that was read, from 2D-array of strings to 1D-array of type byte. It might work if I simply sieve all the array, combine each string value to a Json key.

File: file.go (hunk 14 above)
Comment (suleman01851) [reply file.go]: I have implemented it, now it will work.

File: file.go (hunk 16 above)
Comment (suleman01851) [reply file.go]: sure I will rectify it

File: file.go (hunk 18 above)
Comment (suleman01851) [reply file.go]: The smallest accessible memory unit in GO is byte/uint8, and also data returned after parsing json is also of type byte.

File: file.go (hunk 20 above)
Comment (suleman01851) [reply file.go]: If you are talking about '}' followed by "else", GO's compiler prompts syntax error on this.

File: file.go (hunk 21 above)
Comment (suleman01851) [reply file.go]: Yes, I didn't design it that way. After reading "return the file data in CSV format", I thought it would return data,read from CSV file, in form of [ ] byte. Now, I've edited it, and it will convert data, read from Json file, to CSV.

File: jcsv.go (hunk 22 above)
Comment (suleman01851) [reply jcsv.go]: I am not sure, why it is written that way, I guess it might be prompting error that time.

File: jcsv.go (hunk 23 above)
Comment (suleman01851) [reply jcsv.go]: Sure, in next push I will comment complete code.

File: file.go (hunk 20 above)
Comment (suleman01851) [reply file.go]: I tried move '{' to next line, and when I save code in VS code, it automatically moves the '{' (that follows "else") on the previous line.

File: file.go (hunk 19 above)
Comment (suleman01851) [reply file.go]: Yes, it works, it is interesting to observe that it calls a copy constructor, but if we try to do
"returnFile.csvHeader[j]=readBuffer[0][j]", it does not work because out of range indexing.

File: file.go [hunk 24]
```diff
@@ -1,37 +1,145 @@
...
+	var fileObj file
+	fileStream, fileOpenError := os.Open(path)
+
+	if fileOpenError != nil {
+		return file{}, fileOpenError
+	}
+
+	var fileReadError error
+	defer fileStream.Close()
+	fileObj.jsonData, fileReadError = ioutil.ReadAll(fileStream) // it reads data from json file and return in []byte/uint8 format
```
Comment (IamFaizanKhalid) [inline file.go]: Why not use `[]byte` then?

Comment (IamFaizanKhalid) [inline file.go]: You can use `ParseOpenedJsonFile()` to do the rest.

File: file.go [hunk 25]
```diff
@@ -1,37 +1,145 @@
...
+	}
+
+	//all the data has been transformed from [][]string to [] byte type, but still in CSV format
+	returnedData, err := CsvToJson(dataToBeReturned, true) // this function call will convert it to json
+	if err == nil {
+		return returnedData
+	}
+
 	return nil
 }
```
Comment (IamFaizanKhalid) [inline file.go]: If the read file was JSON, it will not return proper JSON.

File: go.mod [hunk 26]
```diff
@@ -1,3 +0,0 @@
-module github.com/datumbrain/jcsv
-
-go 1.16
```
Comment (IamFaizanKhalid) [inline go.mod:3]: Why is this file deleted?

**Review by IamFaizanKhalid — CHANGES_REQUESTED** [review]

File: go.mod (hunk 26 above)
Comment (suleman01851) [reply go.mod:3]: I had to create my own mod file, I could not execute my main function without doing so.

File: file.go (hunk 24 above)
Comment (suleman01851) [reply file.go]: Yes, that is a good approach of reusability, but I wrote this function after writing ParseJsonFile().

Comment (suleman01851) [reply file.go]: Both are 8-bit long, at that time, I didn't pay heed to their differences.

File: file.go [hunk 27]
```diff
@@ -1,37 +1,140 @@
...
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+)
 
 type file struct {
-	// TODO: define variables if needed
+	data []map[string]interface{} //a 1D-array of maps having string keys and empty interface values
```
Comment (IamFaizanKhalid) [inline file.go:13]:
```suggestion
	// to store parsed data
	data []map[string]interface{}
```

File: file.go [hunk 28]
```diff
@@ -1,37 +1,140 @@
...
 func ParseJsonFile(path string) (file, error) {
-	// TODO: open and read the given file into your `file` object
-	return file{}, nil
+	jsonFile, openError := os.Open(path)
+	if openError != nil {
+		return file{}, openError
+	}
+
+	defer jsonFile.Close()
+	return ParseOpenedJsonFile(jsonFile)
```
Comment (IamFaizanKhalid) [inline file.go:23]:
```suggestion
	defer jsonFile.Close()
	
	return ParseOpenedJsonFile(jsonFile)
```

File: file.go [hunk 29]
```diff
@@ -1,37 +1,140 @@
...
 }
 
 func ParseOpenedJsonFile(f *os.File) (file, error) {
-	// TODO: read the given file into your `file` object
-	return file{}, nil
+	byteValue, readError := ioutil.ReadAll(f) //ReadAll gives read data in []byte
```
Comment (IamFaizanKhalid) [inline file.go:27]: Put all comments above the code.
Please refer to [effective go](https://golang.org/doc/effective_go#commentary) when you need to know about formatting conventions. Note that these are just conventions not hard rules.

File: file.go [hunk 30]
```diff
@@ -1,37 +1,140 @@
...
-	// TODO: open and read the given file into your `file` object
-	return file{}, nil
+	csvFile, openErr := os.Open(path)
+
+	if openErr != nil {
+		return file{}, openErr
+	}
+
+	defer csvFile.Close()
+	return ParseOpenedCsvFile(csvFile, hasHeaders)
```
Comment (IamFaizanKhalid) [inline file.go:58]:
```suggestion
	defer csvFile.Close()
	
	return ParseOpenedCsvFile(csvFile, hasHeaders)
```

**Review by IamFaizanKhalid — CHANGES_REQUESTED** [review]

File: file.go (hunk 29 above)
Comment (suleman01851) [reply file.go:27]: Sure, I will read this article.

File: file.go (hunk 27 above)
Comment (suleman01851) [reply file.go:13]: Sure Faizan bhai, I will follow these paradigms of commenting and writing code.
